internal/datasources/service/service.go

```diff
@@ -0,0 +1,207 @@
+	// GetByName returns a data source by name.
+	GetByName(ctx context.Context, name string, project uuid.UUID, opts *ReadOptions) (*minderv1.DataSource, error)
+
+	// GetByID returns a data source by ID.
+	GetByID(ctx context.Context, id uuid.UUID, project uuid.UUID, opts *ReadOptions) (*minderv1.DataSource, error)
+
+	// List lists all data sources in the given project.
+	List(ctx context.Context, project uuid.UUID, opts *ReadOptions) ([]*minderv1.DataSource, error)
+
+	// Create creates a new data source.
+	Create(ctx context.Context, ds *minderv1.DataSource, opts *Options) (*minderv1.DataSource, error)
+
+	// Update updates an existing data source.
+	Update(ctx context.Context, ds *minderv1.DataSource, opts *Options) (*minderv1.DataSource, error)
+
+	// Delete deletes a data source in the given project.
+	//
+	// Note that one cannot delete a data source that is in use by a rule type.
+	Delete(ctx context.Context, id uuid.UUID, project uuid.UUID, opts *Options) error
+
+	// ValidateRuleTypeReferences takes the data source declarations in
+	// a rule type and validates that the data sources are available
+	// in the project hierarchy.
+	//
+	// Note that the rule type already contains project information.
+	ValidateRuleTypeReferences(ctx context.Context, rt *minderv1.RuleType, opts *Options) error
+
+	// BuildDataSourceRegistry bundles up all data sources referenced in the rule type
+	// into a registry.
+	BuildDataSourceRegistry(ctx context.Context, rt *minderv1.RuleType, opts *Options) (*v1datasources.DataSourceRegistry, error)
+}
+
+type dataSourceService struct {
+	store db.Store
+
+	// This is a function that will begin a transaction for the service.
+	// We make this a function so that we can mock it in tests.
+	txBuilder func(d *dataSourceService, opts txGetter) (serviceTX, error)
+}
+
+// NewDataSourceService creates a new data source service.
+func NewDataSourceService(store db.Store) *dataSourceService {
+	return &dataSourceService{
+		store:     store,
+		txBuilder: beginTx,
+	}
+}
+
+// WithTransactionBuilder sets the transaction builder for the data source service.
+//
+// Note this is mostly just useful for testing.
+func (d *dataSourceService) WithTransactionBuilder(txBuilder func(d *dataSourceService, opts txGetter) (serviceTX, error)) {
+	d.txBuilder = txBuilder
+}
+
+// Ensure that dataSourceService implements DataSourcesService.
+var _ DataSourcesService = (*dataSourceService)(nil)
+
+func (d *dataSourceService) GetByName(
+	ctx context.Context, name string, project uuid.UUID, opts *ReadOptions) (*minderv1.DataSource, error) {
+	return d.getDataSourceSomehow(
+		ctx, project, opts, func(ctx context.Context, tx db.ExtendQuerier, projs []uuid.UUID,
+		) (db.DataSource, error) {
+			ds, err := tx.GetDataSourceByName(ctx, db.GetDataSourceByNameParams{
+				Name:     name,
+				Projects: projs,
+			})
+			if err != nil {
+				if errors.Is(err, sql.ErrNoRows) {
+					return db.DataSource{}, util.UserVisibleError(codes.NotFound,
```

Comment:
Isn't it better to return nil instead of empty object in case of errors? 

_Comment is valid for the other use cases below_

Reply:
this sub-function returns an object and not a pointer, since that's the type of object that the database wrappers usually returns. we do return pointers and thus `nil` underneath though.